qa/vagrant/src/test/resources/packaging/scripts/90_reinstall.bats

```diff
@@ -0,0 +1,88 @@
+# dramatically change your system. It removes the 'elasticsearch'
+# user/group and also many directories. Do not execute this file
+# unless you know exactly what you are doing.
+
+# The test case can be executed with the Bash Automated
+# Testing System tool available at https://github.com/sstephenson/bats
+# Thanks to Sam Stephenson!
+
+# Licensed to Elasticsearch under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# Load test utilities
+load packaging_test_utils
+load os_package
+
+# Cleans everything for the 1st execution
+setup() {
+    skip_not_dpkg_or_rpm
+}
+
+@test "[REINSTALL] install" {
+    clean_before_test
+    install_package
+}
+
+@test "[REINSTALL] start elasticsearch" {
+    # edit the config file so that removing via the rpm does not destroy
+    # /etc/elasticsearch
+    echo "cluster.name: reinstall" >> /etc/elasticsearch/elasticsearch.yml
+    # edit the systemd elasticsearch.service so that removal does not
+    # destroy it
+    is_systemd && perl -p -i -e 's/#LimitMEMLOCK=infinity/LimitMEMLOCK=infinity/g' /usr/lib/systemd/system/elasticsearch.service
+    # so that some files are written to data and logs
+    start_elasticsearch_service
+}
+
+@test "[REINSTALL] stop elasticsearch" {
+    stop_elasticsearch_service
+}
+
+@test "[REINSTALL] purge elasticsearch" {
+    purge_elasticsearch
+}
+
+@test "[REINSTALL] verify preservation" {
+    [ -d "/var/lib/elasticsearch/" ]
+    is_dpkg && [ -d "/var/log/elasticsearch" ]
+    [ -d "/etc/elasticsearch" ]
+    [ -f "/etc/elasticsearch/elasticsearch.yml.rpmsave" ]
+    is_systemd && [[ -f "/usr/lib/systemd/system/elasticsearch.service" || -f "/usr/lib/systemd/system/elasticsearch.service.rpmsave" ]]
+}
+
+@test "[REINSTALL] chown directories" {
+    sudo chown -R root:root /var/lib/elasticsearch
+    [ -d "/var/log/elasticsearch" ] && sudo chown -R root:root /var/log/elasticsearch
+    sudo chown -R root:root /etc/elasticsearch
+}
+
+@test "[REINSTALL] reinstall elasticsearch" {
+    install_package
+}
+
+@test "[REINSTALL] check ownership" {
+    assert_recursive_ownership /var/lib/elasticsearch elasticsearch elasticsearch
+    assert_recursive_ownership /var/log/elasticsearch elasticsearch elasticsearch
+    assert_recursive_ownership /etc/elasticsearch root elasticsearch
+}
```

Comment:
I've reverted these changes; there are some behavior changes between RPM versions that I think require more careful investigation in a separate pull request.